### PR TITLE
Add validation for MetadataType attribute - Issue #46678

### LINKED ISSUE #46678: MetadataTypeAttribute doesn't influence DataAnnotations validation result
### Update by @SteveSandersonMS: please see the comment https://github.com/dotnet/runtime/issues/46678#issuecomment-747533844 for a description of the underlying issue/question.

Please ignore the parts of this thread that are specific to Blazor. That's just how the issue was originally reported, but it looks like the underlying concern is about DataAnnotations itself.

---

### Describe the bug
When I use the MetadataType attribute (from System.ComponentModel.DataAnnotations) on the class for the purpose of validation I get the following error in the console window
`WASM: System.TypeLoadException: Could not resolve type with token 01000027 from typeref (expected class 'System.ComponentModel.DataAnnotations.MetadataTypeAttribute' in assembly 'System.ComponentModel.Annotations, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a')`

Greater details of the error are shown in this file...[console screenshot - validation error](https://user-images.githubusercontent.com/5938313/79199691-42595400-7e2d-11ea-95bc-3c2248d8b60e.png

Please note my sample application does works when I put all validation attributes into main the class.

For clarity:
This approach works...

```
public partial class PersonViewModel
{
    [Required(ErrorMessage = "First name is required.")]
    public string Fname { get; set; }
}
```

This approach fails...

```
public partial class PersonViewModel
{
    public string Fname { get; set; }
}

[MetadataType(typeof(PersonViewModelMetaData))]
public partial class PersonViewModel
{
}

public class PersonViewModelMetaData
{
    (ErrorMessage = "First name is required.")]
    public string Fname { get; set; }
}
```

### To Reproduce
Please run the attached sample blazor application. To see the error please click the navigation link labelled "validation - failure". On the page, just simply press the submit button to see the error.
[BlazorAppValidation.zip](https://github.com/dotnet/aspnetcore/files/4474042/BlazorAppValidation.zip)

### Further technical details
- Include the output of `dotnet --info`
```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.300-preview-015048
 Commit:    13f19b4682

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.300-preview-015048\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

.NET Core SDKs installed:
  3.1.101 [C:\Program Files\dotnet\sdk]
  3.1.201 [C:\Program Files\dotnet\sdk]
  3.1.300-preview-015048 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

```

### PR DESCRIPTION
**Pull request by Xaeco**
If a MetadataType attribute is applied to a class the associated metadata class is now validated in .NET Core. 

In .NET Framework 4.8 and older versions a similar behavior already exists if you use associate a type and handler like so: 
`TypeDescriptor.AddProviderTransparent(
                new AssociatedMetadataTypeTypeDescriptionProvider(typeof(Person), typeof(PersonMetadata)),
                typeof(Person));` 

To do this the System.ComponentModel.DataAnnotations.**ValidationAttributeStore** class was altered to check for a MetadataType attribute. If found it inspects the associated class and merges any validation attributes with any from the class being validated.

Added tests to the System.ComponentModel.DataAnnotations.**Validator** for as many scenarios as I could think of.

Fixes #46678

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=51772) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=51772) before we can accept your contribution.<br/><br/>:x: Xaeco [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=51772)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=51772) it.</sub>

**Comment by Xaeco** [comment]
@ajcvickers what is the process and timeline to get someone to review my PR? Also I can't get this to build due to the Helix failures that keep happening. Seems to be a common issue.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:239]: This is not a valid justification for suppressing the trim warnings. The `Type` being passed into the ctor is not the same type that is being reflected upon here.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:243]: Can you refactor 

[https://github.com/dotnet/runtime/blob/51d6f177e14c21eb4511b3e5a1d6fb61d90da61d/src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs#L129-L133](https://github.com/dotnet/runtime/blob/51d6f177e14c21eb4511b3e5a1d6fb61d90da61d/src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs#L129-L133)

and reuse the same logic here? That way we are consistent in how we are reading this attribute.

Once you do that, the trim warnings will go away because the Type returned from that refactored method can be annotated with `DynamicallyAccessedMemberTypes.All`, which means you can reflect over the Properties on that type.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:262]: I don't believe you need a `!` here on `baseProperty!`. The dictionary doesn't contain `null`s, and `TryGetValue` returned `true`.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:239]: Thank you. Working on it 😊

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:243]: Thanks. I refactored it (not pushed yet) but it's complaining about possible nullability. I'll merge master in and see if it's a transitory issue.

Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Using `(MetadataTypeAttribute)Attribute.GetCustomAttribute` I was getting:
`Error	CS8600	Converting null literal or possible null value to non-nullable type.`. 

I noticed in file [AssociatedMetadataTypeTypeDescriptor.cs](https://github.com/dotnet/runtime/blob/51d6f177e14c21eb4511b3e5a1d6fb61d90da61d/src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/AssociatedMetadataTypeTypeDescriptor.cs#L129-L133) it has `#nullable disable` which is suppressing the error.

I switched to `(MetadataTypeAttribute)Attribute.GetCustomAttributes(_type, typeof(MetadataTypeAttribute)).First()`

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: VS used to complain it was potentially null. It doesn't now. Thanks

**Comment by Xaeco** [comment]
@ajcvickers @eerhardt @lajones when's the next time this can be reviewed ?
Thanks

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:238]:
```suggestion
```

This attribute isn't valid. `DynamicallyAccessedMembers` are only valid to annotate a `Type` or a `string` (which is a type name). Or it can also be applied to a `class` as well.

Here you are saying the `return` value should keep `All` members, but the return value is a `Dictionary<string, PropertyStoreItem>`.

I'm surprised the linker doesn't complain about this (cc @vitek-karas @sbomer).

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:246]:
```suggestion
                    Type t = attribute.MetadataClassType;
```

There's no reason to use `TypeInfo` anymore. This class was an attempt at a redesign of Reflection during Silverlight and early .NET Core days. But with .NET Core 2.0 all the APIs on Type are back. Just use `Type`.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:251]:
```suggestion
                    foreach (PropertyInfo property in t.GetProperties())
                    {
                        if (property.GetIndexParameters().Length == 0)
                        {
```

Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:251]: Should you also be filtering `static` properties? `GetProperties()` returns both instance and static properties.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (vitek-karas) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:238]: Good catch @eerhardt - filed an issue in the linker: https://github.com/mono/linker/issues/2065

Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Yes I had initially pasted that in while testing and forgot to check if it was actually required.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Good to know. Corrected. Thanks.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: I've incorporated your suggested code. Is the reasoning of the style change to reduce memory and object iteration for performance reasons?

Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: @eerhardt Re `static` properties: Only public static properties will be processed. This is the same behavior as in .Net Framework. My branch finally got through all the pipelines, yay, progress. Thanks for all your help.

Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: >  Is the reasoning of the style change to reduce memory and object iteration for performance reasons?

Yes. In general, we have been moving away from using Linq in the BCL. See #44923 #44825 #47496 #47873 #44734

Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: I would hope the area owners would review this change. I'm just here reviewing because I was pinged on the issue, and was helping out on the "trim compatibility" parts of it.

@ajcvickers

Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Thanks. Would it be appropriate for you to assign the review to ajcvickers?

**Comment by Xaeco** [comment]
@ajcvickers can we progress this? I and other have a lot of work with this as a dependency.

**Comment by Xaeco** [comment]
Hi @eerhardt. Trying to progress this before .Net 6 is released 😂. I was wondering if this PR is setup/tagged correctly for the associated bug? Also any other settings, or advice to progress this would be appreciated. Thanks

**Comment by ajcvickers** [comment]
Sorry for being so slow on this. We are normally very hesitant to take changes in this area, especially where the behavior in .NET Framework and .NET Core would diverge. However, in this case we are bringing an existing .NET Framework behavior to .NET Core, which actually helps convergence.

@bricelam I think you have some knowledge in this area. Does the change make sense to you?
@roji Another set of eyes would be good.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (roji) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:220]: Can you please remove this newline?

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (roji) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:241]: I'm not sure, but shouldn't this go through AssociatedMetadataTypeTypeDescriptor? Even if not, it should probably go through TypeDescriptorCache.GetAssociatedMetadataType.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (roji) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:253]:
```suggestion
```

**Review by roji — COMMENTED** [review]
As @ajcvickers wrote, apologies for being so slow on this (I've been on vacation in the last three weeks).

I'm unfamiliar with this area, but the main question that comes to mind, is whether we shouldn't simply match the .NET Framework behavior here - more or less by copying the code across - rather than adding the new feature here (as @ajcvickers wrote this area is archived to a large extent). If I understand the situation correctly, this would mean that [MetadataType] wouldn't be respected for validation (it wasn't for .NET Framework), but TypeDescriptor.AddProviderTransparent would.

In any case, deferring to @bricelam (and @ajcvickers) who know this better.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:241]: On .NET Framework - `CreatePropertyStoreItems()` called `TypeDescriptor.GetProperties`, and then filtered the attributes that were explicitly specified. See

https://referencesource.microsoft.com/#System.ComponentModel.DataAnnotations/DataAnnotations/ValidationAttributeStore.cs,211

**Comment by bricelam** [comment]
+1 for using the .NET Framework code as reference. When all this code came into .NET Core, `MetadataTypeAttribute` was deliberately excluded (no idea why), and it was later brought in as part of the push to make .NET Framework apps compatible with .NET Core.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Done. Thanks

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Done. Thanks

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:240]: Yes copying the .NET Framework code over works as long as you also use 
`TypeDescriptor.AddProviderTransparent( new AssociatedMetadataTypeTypeDescriptionProvider(typeof(Person), typeof(PersonMetadata)), typeof(Person));`

There was commentary in the code and elsewhere in forums (I can't remember where) indicating that the TypeDescriptor approach was not the way forward, hence this PR. 

Thanks again to @eerhardt for his help.

**Comment by Xaeco** [comment]
@bricelam . So I'm wondering what the next steps are. I'd like to keep this moving.

Appreciate everyone's feedback, thanks.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (roji) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:240]: You may be referring to https://github.com/dotnet/runtime/issues/46678#issuecomment-787738854. However, as @bricelam and I wrote, we're looking to align the .NET Core implementation here with the .NET Framework, by copying the code across from the .NET Framework reference sources, rather than writing new support here as you've done (if I understand everything correctly, that corresponds to solution 1 in the linked comment).

**Comment by roji** [comment]
See https://github.com/dotnet/runtime/pull/51772#discussion_r671079671

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:240]: Shall I copy that over? Or are there other things you want to check/consider first?

Comment (roji) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:240]: I think copying that over and thereby aligning the .NET Core implementation to the .NET Framework one is what we're looking for - I don't think there's something else to check.

Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: The build is failing due to trim analysis errors. I've tried for a while but I can't figure out how to suppress them. Any pointers appreciated.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (vitek-karas) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:240]: Why is this method public? It seems to only be used by this class.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (vitek-karas) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: The annotation on the `_type` field doesn't match the required annotation on `TypeDescriptor.GetProperties`. Unfortunately we had to use `DynamicallyAccessedMemberTypes.All` annotation on the `GetProperties` (since under the hood it creates a real type descriptor which needs to know pretty much everything about the type).
To fix the first warning (IL2077) change the `TypeStoreItem.DynamicallyAccessedTypes` to `DynamicallyAccessedMemberTypes.All`.

To fix the second one (IL2026) it will be necessary to add `RequiresUnreferencedCode` attribute onto the calling method (`CreatePropertyStoreItems`). That will in turn force you to add that attribute to all its callers and so on. Fortunately it seems that all the caller up one level already have the RUC attribute on them, so the change should be relatively simple.

The last warning (IL2072) will need to be suppressed by adding the `RequiresUnreferencedCode` attribute onto the `GetExplicitAttribute`. Reasoning is the same as the RUC On TypeDescriptor.GetProperties.

Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Thanks @vitek-karas. I was avoiding changing the annotation on the `_type` field for some reason. I should have a PR in shortly.

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:240]: 🤔 that would be a copy & paste fail from .NET Framework. Changed to internal. Thanks

Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: (nit) Does it even need to be `internal`? Why not `private`?

File: src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs:218]: It's not `_type` that is the problem. It's the types of its properties.

```suggestion
            [RequiresUnreferencedCode("The Types of _type's properties cannot be statically discovered.")]
```

File: src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/ValidatorTests.cs
Comment (eerhardt) [inline src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/ValidatorTests.cs:368]: Can this be reverted? It isn't necessary and will only cause unnecessary churn in source history.

Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/ValidatorTests.cs:368]: What happened to the new tests that you added previously? I think fixing this bug should introduce some new tests that ensure the bug is fixed.

**Comment by karelz** [comment]
Http3_MsQuic test failures are tracked in #56090. Should be fixed and merged shortly. Sorry for the inconvenience!

File: src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/ValidatorTests.cs
Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/ValidatorTests.cs:368]: > Can this be reverted? It isn't necessary and will only cause unnecessary churn in source history.

Do you mean revert the removal of the tests?

Comment (Xaeco) [reply src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/ValidatorTests.cs:368]: > What happened to the new tests that you added previously? I think fixing this bug should introduce some new tests that ensure the bug is fixed.

I'll adjust the tests to include the new behavior.

Comment (eerhardt) [reply src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/ValidatorTests.cs:368]: > Do you mean revert the removal of the tests?

At first I meant "Can this unnecessary blank line be reverted?". Adding unnecessary blank lines is noise in a PR.

But then I realized that we are no longer adding unit tests in this PR. We should always try to add unit tests for bug fixes, to ensure the behavior is correct and doesn't regress in the future.

> I'll adjust the tests to include the new behavior.

Thank you.

**Comment by eerhardt** [comment]
This looks good to me. Thanks @Xaeco for the contribution here.

@roji @bricelam - any more feedback?

**Review by eerhardt — APPROVED** [review]

**Comment by Xaeco** [comment]
Awesome it's approved! 😀. Thanks everyone. This is my first GitHub PR. Does it get auto-completed & merged or do I need to do something?

**Comment by ajcvickers** [comment]
@bricelam Can you take a final look at this when you get in today?

@danmoseley @Pilchie Does this need to go through Ask Mode to get in for .NET 6?

**Comment by danmoseley** [comment]
@ajcvickers there are no permissions (ever?) needed to fix anything in dotnet/runtime main. However on Aug 17 it will become a 7.0 branch so there would be ceremony if it wasn't in by then.

**Comment by bricelam** [comment]
@Xaeco It looks like `GetTypeStoreItem` should also use `TypeDescriptor` to pick up type-level attributes on the buddy type. Wanna make the update and add some tests for this scenario?

**Comment by bricelam** [comment]
Updated to honor class-level attributes. Squashed and rebased on top of `main`.

**Review by bricelam — APPROVED** [review]

**Comment by bricelam** [comment]
Looks like scenarios involving [dynamically added properties](https://putridparrot.com/blog/dynamically-extending-an-objects-properties-using-typedescriptor/) will also be enabled by this PR.

**Comment by Xaeco** [comment]
> @Xaeco It looks like `GetTypeStoreItem` should also use `TypeDescriptor` to pick up type-level attributes on the buddy type. ~Wanna make the update and add some tests for this scenario?~

@bricelam thanks for making those changes. I started to but got interrupted and should've mentioned I was working on it.

**Comment by IanKemp** [comment]
Thanks guys!

**Comment by bricelam** [comment]
And thank you, community, for making it happen! Things like this that bring parity with .NET Framework make everyone's migration go a little smoother.

**Comment by Gambero81** [comment]
Glad to find this issue merged in .net6, can ported to .net5?

**Comment by roji** [comment]
@Gambero81 this is very unlikely to be backported to .NET 5.0; patches are only done for serious bugs for which no workaround is available.

**Pull request merged**
